File: Java/Algorithms/Geometrical Algorithms/LinkedList.java
//Floyd’s Cycle-Finding Algorithm:- To find a loop in linked list


public class LinkedList {
    Node head;
    class Node {		//creation of a node
        int data;
        Node next;
        Node(int d)
        {
            data = d;
            next = null;
        }
    }
    public void push(int new_data)	//insertion of a new node
    {
        Node new_node = new Node(new_data);
        new_node.next = head;
        head = new_node;
    }
 
    void detectLoop()			//floyd's cycle finding algorithm using two pointers
    {
        Node slow_p = head, fast_p = head;
        int flag = 0;
        while (slow_p != null && fast_p != null
               && fast_p.next != null) {
            slow_p = slow_p.next;
            fast_p = fast_p.next.next;
            if (slow_p == fast_p) {		//if two pointers are same then there is a loop
                flag = 1;
                break;
            }
        }
        if (flag == 1)
            System.out.println("Loop found");
        else
            System.out.println("Loop not found");
    }
    public static void main(String args[])		//main method
    {
        LinkedList llist = new LinkedList();
 
        llist.push(20);
        llist.push(4);
        llist.push(15);
        llist.push(10);
        llist.head.next.next.next.next = llist.head;
 
        llist.detectLoop();
    }
}

//Time Complexity - O(n)
//Space Complexity - O(1)

File: Javascript/Data-structures/Linked-List/linkedList.js
class Node {
  constructor(data, next = null) {
    this.data = data;
    this.next = next;
  }
}

class LinkedList {
  constructor(head = null) {
    this.head = head;
  }

  insert(node) {
    if (this.head === null) {
      this.head = node;
      return;
    }
    let curr = this.head;
    while (curr.next) {
      curr = curr.next;
    }
    curr.next = node;
  }

  display() {
    let curr = this.head;
    while (curr) {
      console.log(curr.data);
      curr = curr.next;
    }
  }

  deleteNode(data) {
    let curr = this.head;
    if (curr && curr.data === data) {
      this.head = curr.next;
      return;
    }
    while (curr.next) {
      if (curr.next.data === data) {
        curr.next = curr.next.next;
        return;
      }
      curr = curr.next;
    }
  }
}

let arr = [1, 2, 3, 4, 5, 6];

const ll = new LinkedList();

for (let i of arr) {
  const newNode = new Node(i);
  ll.insert(newNode);
}

ll.display();
ll.deleteNode(1);
ll.display();
